src/btree/bt_sync.c

```diff
@@ -109,44 +109,53 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_ADDR *addr;
     WT_CELL_UNPACK vpack;
+    WT_PAGE_MODIFY *mod;
 
     /* Ignore root pages as they can never be deleted. */
     if (__wt_ref_is_root(ref))
         return (false);
 
-    /* Guard against marking an in-memory page to be deleted. */
-    if (ref->state != WT_REF_DISK)
-        return (false);
-
     /* Ignore internal pages, these are taken care of during reconciliation. */
     if (!__wt_ref_is_leaf(session, ref))
         return (false);
 
+    /* In-memory ref address must be from an off page address. */
+    mod = (ref->page != NULL) ? (ref->page->modify) : NULL;
+    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
+        return (__wt_txn_visible_all(
+          session, mod->mod_replace.newest_stop_txn, mod->mod_replace.newest_stop_ts));
+
+    if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK)
+        return (__wt_txn_visible_all(
+          session, mod->mod_multi_newest_stop_txn, mod->mod_multi_newest_stop_ts));
+
```

Comment:
It seems confusing to me we use the `mod_replace` structure directly in this case, but we've aggregated the information for the `mod_multi` structure. Why are we handling them differently?

Reply:
When the page is reconciled into multiple blocks, Instead of checking each individual blocks to decide the obsoleteness, we aggregated them, but whereas for `mod_replace` scenario, it is just one block, I feel directly referring it may be enough.

Reply:
My approach would be to aggregate when doing the work in checkpoint, rather than modifying reconciliation and adding new fields to the `WT_PAGE_MODIFY` structure.

Reply:
Fixed.